### DISCUSSION TITLE
DrawText() results in "wobbly" drawn text?

### DISCUSSION THREAD

**Discussion by QuantumToasted** [SixLabors/ImageSharp.Drawing#336] [Q&A] [answered]
I don't know how else to describe this, but I'm trying to isolate a strange issue occurring in a much larger section of image-generating code. 
![image](https://github.com/user-attachments/assets/ef493706-c7c3-4b2b-8fbc-6f4e883729d7)

Note that **all** the text here has a certain..."wobbly-ness" to it. The individual letters (glyphs?) are very poorly vertically aligned with themselves. I am not specifying the text be drawn along any path...as far as I can tell. The font is `Victor Mono`, the same font I use in my IDE, so I know it is not the way the font is designed to look.
All my text is drawn the same way (with the only difference being the Origin and WrappingLength properties), so here is one example, the text located at the bottom (tribute name):
```cs
x.DrawText(new RichTextOptions(Font.CreateFont(TRIBUTE_NAME_FONT_SIZE))
    {
        HorizontalAlignment = HorizontalAlignment.Center,
        Origin = new Vector2(TRIBUTE_AVATAR_BOUNDS / 2f, DISTRICT_HEADER_OFFSET + TRIBUTE_AVATAR_BOUNDS + TRIBUTE_STATUS_OFFSET + VERTICAL_PAD * 3),
        WrappingLength = TRIBUTE_AVATAR_BOUNDS,
        TextAlignment = TextAlignment.Center
    },
    tribute.Name,
    Brushes.Solid(Color.White), 
    Pens.Solid(Color.White));
```
While it is not directly related to the issue (I hope), is there a way to draw text using RichTextOptions but **NOT** having to specify both a Brush and a Pen? I would prefer just a simple "render all this text the same exact color with no effects or gradients".

Library version: `2.1.4`

**Comment by QuantumToasted** [comment]
Update: It appears to be some sort of issue related to rendering fonts at...lower resolutions?
Completely unrelated to attempting to fix the issue on my end, I doubled all size values by 2, including fonts, and it appears to be rendering more correctly now?

These images are not identical due to the nature of how they are generated, but as you can observe...
Before (original scale):
![status-1](https://github.com/user-attachments/assets/602f9985-e06e-47f4-adfb-b356976a0314)
After (2x scale):
![status](https://github.com/user-attachments/assets/9eb080af-cd22-44ff-8c96-20792ee3c8e2)

Text appears to render perfectly fine and not misaligned at all.

**Reply by JimBobSquarePants** [reply]
You’ll need to give us a reproducible sample to work with. There’s no way to investigate this without actual things like the font size etc

**Reply by QuantumToasted** [reply]
Thanks. I thought to ask the simpler question first with a small snippet in case it was something obvious I was missing. I was migrating code from 2020 or even earlier (including an I# version from 2020 or earlier) and the syntax has clearly changed immensely.

As this generation code is quite lengthy, here's a [gist](https://gist.github.com/QuantumToasted/ac98be5cf8bb78762ce396f29c9712dc). I've included really simple mocks of the classes that are supposed to go in there, at least as they are used in this service. If you need like...actual real world sample data, I can provide, but the issue is with the text, not images.
The font I'm using is [Victor Mono](https://rubjo.github.io/victor-mono/), and when SCALE was set to 1, the font sizes were `15f` for the district names and `11.25f` for the tribute status/name.

As a separate thing I noticed after doubling the scale, there seems to be some weird issue with the correctly-aligned text. Like the text outline and fill are misaligned. I'm unsure how well it's visible in this screenshot...
![image](https://github.com/user-attachments/assets/016dd8b5-cfdb-417b-acf5-48a389d05bf1)

If this gist is just too much to try to pick apart yourself, I will try my hand at isolating a small section of code.

**Comment by tocsoft** [comment]
Try setting the pen to `null`, that should skip the outline phase and should make it more consistent, there was a bug in the version you are using that offset out the outline by a small amount that will likely be compounded at smaller sizes.

**Reply by JimBobSquarePants** [reply, marked as answer]
2.1.4 is the latest version which contains your [fix](https://github.com/SixLabors/ImageSharp.Drawing/pull/331) for the offset outline.

When no pen is set the output is fine.

```c#
FontCollection collection = new();
FontFamily family = collection.Add(@"C:\Users\james\Downloads\VictorMonoAll\TTF\VictorMono-Regular.ttf");
Font font = family.CreateFont(11.25f);

RichTextOptions options = new(font);

const string text = "The quick brown fox etc etc";

FontRectangle advance = TextMeasurer.MeasureAdvance(text, options);

using Image<Rgba32> image = new((int)Math.Ceiling(advance.Width), (int)Math.Ceiling(advance.Height), Color.Black);
image.Mutate(i => i.DrawText(options, text, Color.White));

image.SaveAsPng(@"C:\Users\james\Downloads\wobbly.png");

```

![wobbly](https://github.com/user-attachments/assets/b8c0e6e4-4887-4e8a-a551-187dec4c97c5)

When set it look like there's still an issue. 
```c#
FontCollection collection = new();
FontFamily family = collection.Add(@"C:\Users\james\Downloads\VictorMonoAll\TTF\VictorMono-Regular.ttf");
Font font = family.CreateFont(11.25f);

RichTextOptions options = new(font);

const string text = "The quick brown fox etc etc";

FontRectangle advance = TextMeasurer.MeasureAdvance(text, options);

using Image<Rgba32> image = new((int)Math.Ceiling(advance.Width), (int)Math.Ceiling(advance.Height), Color.Black);
image.Mutate(i => i.DrawText(options, text, Brushes.Solid(Color.White), Pens.Solid(Color.White)));

image.SaveAsPng(@"C:\Users\james\Downloads\wobbly-2.png");

```
![wobbly-2](https://github.com/user-attachments/assets/c1042280-e9fc-493b-8502-f4c73f5f80e4)

**Reply by QuantumToasted** [reply]
Thanks, that does seem to be the issue. Maybe my IDE was bugging out and not showing proper XMLdocs, but I didn't have any reason to believe that it was allowed to set `pen` to `null` in `DrawText()`. I'm unsure what the library's compatibility level is with NRT, but in either case, it may be a good idea to document somewhere that one can set `pen` (and `brush` too, if so?) to `null`.

In either case, this seems to resolve the issue for me well enough, as I don't really need a pen/outline at all (even if it would make the text appear thinner?). Thank you both!

**Reply by JimBobSquarePants** [reply]
Here’s the API docs for DrawText. There’s several overloads and as demonstrated in my code sample you don’t need to set either a brush or pen.

https://docs.sixlabors.com/api/ImageSharp.Drawing/SixLabors.ImageSharp.Drawing.Processing.DrawTextExtensions.html

We’ll need to raise an issue regarding the offsetting.
